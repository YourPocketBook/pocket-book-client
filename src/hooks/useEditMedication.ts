import isValid from "date-fns/isValid";
import parseISO from "date-fns/parseISO";
import { useReducer } from "react";
import {
  IEditMedication,
  IEditMedicationValid
} from "../medications/medication-form";

export function useEditMedication() {
  const [state, dispatch] = useReducer(
    (
      s: IEditMedication & { valid: IEditMedicationValid },
      a: { prop: keyof IEditMedication } & (
        | { op: "update"; value: string | undefined | null }
        | {
            op: "invalidate";
          }
        | { op: "setValid"; value: boolean }
      )
    ) => {
      if (s.hasOwnProperty(a.prop)) {
        const newData = { ...s };

        switch (a.op) {
          case "update":
            newData[a.prop] = a.value || "";
            if (a.prop === "policyDate") {
              newData.valid[a.prop] = !!a.value && isValid(parseISO(a.value));
            } else {
              newData.valid[a.prop] = !!a.value && a.value.length > 0;
            }
            break;
          case "invalidate":
            newData.valid[a.prop] = false;
            break;
          case "setValid":
            newData.valid[a.prop] = a.value;
            break;
        }
        return newData;
      }
      return s;
    },
    {
      adviceIfDeclined: "",
      adviceIfTaken: "",
      dose: "",
      exclusionCriteria: "",
      form: "",
      inclusionCriteria: "",
      indications: "",
      name: "",
      policyDate: "",
      route: "",
      sideEffects: "",
      valid: {
        adviceIfDeclined: false,
        adviceIfTaken: false,
        dose: false,
        exclusionCriteria: false,
        form: false,
        inclusionCriteria: false,
        indications: false,
        name: false,
        policyDate: false,
        route: false,
        sideEffects: false
      }
    }
  );

  return { state, dispatch };
}
