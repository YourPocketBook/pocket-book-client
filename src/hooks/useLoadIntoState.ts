import { Dispatch, useEffect, useState } from "react";
import { fetcher } from "../fetcher";

/**
 * Fetches JSON from the specified uri and loads each property into individually editable state variables
 *
 * @param uri The URI to load the data from
 * @param setState An object containing setState functions for each property
 * @param [authenticate=false] If set to true, the current token will be passed with the call
 * @returns True if the request is currently running, otherwise false.
 */
export function useLoadIntoEditableState<
  TData extends { [key: string]: string | undefined | null } & object
>(
  uri: string,
  dispatch: Dispatch<{
    op: "update";
    prop: keyof TData;
    value: string | undefined | null;
  }>,
  authenticate = false
): boolean {
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    const fetchAbort = new AbortController();

    const options: RequestInit = {
      cache: "reload",
      method: "GET",
      signal: fetchAbort.signal
    };

    async function func() {
      try {
        const res = authenticate
          ? await fetcher.authFetch(uri, options)
          : await fetch(uri, options);

        if (res.ok) {
          const data = (await res.json()) as TData;
          for (const prop in data) {
            if (data.hasOwnProperty(prop)) {
              dispatch({ op: "update", prop, value: data[prop] });
            }
          }
        }
      } catch (ex) {
        if (!fetchAbort.signal.aborted) {
          // tslint:disable-next-line:no-console
          console.log(`Could not get from ${uri}.`);
        }
      } finally {
        if (!fetchAbort.signal.aborted) {
          setLoading(false);
        }
      }
    }

    func();
    return () => fetchAbort.abort();
  }, [uri, authenticate, dispatch]);

  return loading;
}
