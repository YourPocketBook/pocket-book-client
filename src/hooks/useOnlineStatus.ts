// Included from https://github.com/21kb/react-hooks/blob/master/packages/react-online-status-hook
//
// Copyright Â© 2018 Kunall Banerjee
// MIT License

import { useEffect, useState } from "react";

export function useOnlineStatus() {
  const [state, setState] = useState(navigator.onLine);

  const onOnlineEvent = () => {
    setState(navigator.onLine);
  };

  const onOfflineEvent = () => {
    setState(navigator.onLine);
  };

  useEffect(() => {
    window.addEventListener("online", onOnlineEvent);
    window.addEventListener("offline", onOfflineEvent);

    return () => {
      window.removeEventListener("online", onOnlineEvent);
      window.removeEventListener("offline", onOfflineEvent);
    };
  });

  return state;
}
