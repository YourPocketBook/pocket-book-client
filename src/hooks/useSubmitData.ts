import { useState } from "react";
import { fetcher } from "../fetcher";

type IIsValid<T> = { [P in keyof T]: boolean };
type ISetIsValid<T> = { [P in keyof T]?: (val: boolean) => void };
type ISetCustomValid<T> = { [P in keyof T]?: (val: string) => void };

/**
 * Hook that handles submitting form data and responding to server-side validation.
 *
 * @param value Object that contains the current values to send to the server
 * @param valid Object that contains the validation status of each item in value
 * @param setValid Object that contains a set function for each item in valid
 * @param url The URI the data should be submitted to
 * @param options The options to control the submitting
 */
export function useSubmitData<TData extends object>(
  value: TData,
  valid: IIsValid<TData>,
  setValid: ISetIsValid<TData>,
  url: string,
  options: {
    setShowValidation?: any;
    method?: "PUT" | "POST";
    authenticate?: boolean;
    customVal?: ISetCustomValid<TData>;
    success?(arg?: {}): void;
  } = {}
) {
  const [serverError, setServerError] = useState(false);
  const [notAuthorised, setNotAuthorised] = useState(false);
  const [inProgress, setInProgress] = useState(false);

  async function submit(e: { preventDefault(): void }) {
    e.preventDefault();

    for (const prop in value) {
      if (value.hasOwnProperty(prop) && valid.hasOwnProperty(prop)) {
        if (!valid[prop]) {
          if (options.setShowValidation) {
            options.setShowValidation(true);
          }
          return;
        }
      }
    }

    if (options.setShowValidation) {
      options.setShowValidation(false);
    }

    setServerError(false);
    setNotAuthorised(false);
    setInProgress(true);

    const method = options.method ? options.method : "POST";

    const requestOptions = {
      body: JSON.stringify(value),
      headers: {
        "Content-Type": "application/json"
      },
      method
    };

    try {
      const res = options.authenticate
        ? await fetcher.authFetch(url, requestOptions)
        : await fetch(url, requestOptions);

      if (res.ok) {
        if (options.success) {
          if (res.status !== 204) {
            options.success(await res.json());
          } else {
            options.success();
          }
        }
      } else {
        if (options.setShowValidation) {
          options.setShowValidation(true);
        }

        if (res.status === 400 || res.status === 409 || res.status === 403) {
          const body = (await res.json()) as Array<{
            path: keyof TData;
            error: "isBlank" | "isInvalid";
          }>;

          for (const b of body) {
            if (b.error === "isBlank" || b.error === "isInvalid") {
              const validFunc = setValid[b.path];

              if (validFunc) {
                validFunc(false);
              }
            } else if (options.customVal) {
              const func = options.customVal[b.path];
              if (func) {
                func(b.error);
              }
            }
          }
        } else if (res.status === 401) {
          setNotAuthorised(true);
        } else {
          setServerError(true);
        }
      }
    } finally {
      setInProgress(false);
    }
  }

  return { submit, serverError, notAuthorised, inProgress };
}
