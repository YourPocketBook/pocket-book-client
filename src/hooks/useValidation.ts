import EmailValidator from "email-validator";
import { useState } from "react";

/**
 * Sets up a state variable that is also validated
 *
 * - "text" - No content validation is run on the text
 * - "email" - The value is checked to be a valid email address
 *
 * @param initialValue The initial value of the state variable
 * @param type The type of validation to run on the variable
 * @param required The value must be non-empty
 * @returns The current value, the validation status, the set function for the current value and the set function
 * for the validation status, in that order.
 */
export function useValidation(
  initialValue: string,
  type: "text" | "email"
): [string, boolean, (value: string) => void, (value: boolean) => void] {
  function checkIsValid(val: string) {
    if (val.trim().length === 0) {
      return false;
    }
    if (type === "email") {
      return EmailValidator.validate(val);
    }

    return true;
  }

  const [value, setValue] = useState(initialValue);
  const [valid, setValid] = useState(checkIsValid(initialValue));

  function setValidationValue(val: string) {
    setValid(checkIsValid(val));
    setValue(val);
  }

  return [value, valid, setValidationValue, setValid];
}
