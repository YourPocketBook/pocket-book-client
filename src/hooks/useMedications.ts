import { useEffect, useState } from "react";
import { fetcher } from "../fetcher";
import { IMedication } from "../model/medication";
import { useOnlineStatus } from "./useOnlineStatus";

const medicationsKey = "medications";
const listEndpoint = "/api/get-medications";
const detailEndpoint = "/api/get-medication";

export function useMedicationList() {
  const [isLoading, setIsLoading] = useState(false);
  const [serverError, setServerError] = useState(false);
  const [updateAvailable, setUpdateAvailable] = useState(false);
  const isOnline = useOnlineStatus();

  const localStoreString = localStorage.getItem(medicationsKey);
  const localData = localStoreString
    ? (JSON.parse(localStoreString) as Array<Partial<IMedication>>)
    : [];
  const [medicationList, setMedicationList] = useState(localData);

  useEffect(() => {
    if (!isOnline) {
      return;
    }

    const fetchAbort = new AbortController();

    const options = {
      method: "GET",
      signal: fetchAbort.signal
    };

    setIsLoading(true);

    async function func() {
      try {
        const res = await fetcher.authFetch(listEndpoint, options);

        if (fetchAbort.signal.aborted) {
          return;
        }

        if (res.ok) {
          const data = (await res.json()) as Array<Partial<IMedication>>;

          if (fetchAbort.signal.aborted) {
            return;
          }

          const cacheString = localStorage.getItem(medicationsKey);
          const oldData = cacheString
            ? (JSON.parse(cacheString) as Array<Partial<IMedication>>)
            : [];

          const currentIds = data
            .filter(m => m && m.id)
            .map(m => m.id as number);
          const previousIds = oldData
            .filter(m => m && m.id)
            .map(m => m.id as number);

          const updatedItems = oldData
            .filter(m => currentIds.includes(m.id as number))
            .map(m => {
              const update = data.find(n => n.id === m.id);
              if (update && update.name) {
                m.name = update.name;
              }
              return m;
            })
            .concat(data.filter(m => !previousIds.includes(m.id as number)));

          setMedicationList(updatedItems);
          localStorage.setItem(medicationsKey, JSON.stringify(updatedItems));
          setUpdateAvailable(true);
          setServerError(false);
        } else {
          setServerError(true);
        }
      } catch (ex) {
        if (!fetchAbort.signal.aborted) {
          setServerError(true);
        }
      } finally {
        if (!fetchAbort.signal.aborted) {
          setIsLoading(false);
        }
      }
    }

    func();

    return () => fetchAbort.abort();
  }, [isOnline]);

  return {
    isLoading,
    medications: medicationList
      .filter(m => m && m.id)
      .map(m => ({
        availableOffline: !!m.inclusionCriteria,
        id: m.id as number,
        name: m.name as string
      })),
    serverError,
    updateAvailable
  };
}

export function useMedication(id: string, refresh: boolean) {
  const [isLoading, setIsLoading] = useState(false);
  const [updateAvailable, setUpdateAvailable] = useState(false);
  const [loadFailed, setLoadFailed] = useState(false);
  const isOnline = useOnlineStatus();

  const localStoreString = localStorage.getItem(medicationsKey);
  const data = localStoreString
    ? (JSON.parse(localStoreString) as Array<Partial<IMedication>>)
    : undefined;
  const med = data ? data.find(m => m.id === parseInt(id, 10)) : undefined;
  const [medication, setMedication] = useState(med);

  useEffect(() => {
    if (!isOnline) {
      return;
    }

    const fetchAbort = new AbortController();

    const options = {
      cache: refresh ? ("no-cache" as "no-cache") : undefined,
      method: "GET",
      signal: fetchAbort.signal
    };

    setIsLoading(true);
    setLoadFailed(false);

    async function func() {
      try {
        const res = await fetcher.authFetch(`${detailEndpoint}/${id}`, options);

        if (fetchAbort.signal.aborted) {
          return;
        }

        if (res.ok) {
          const newData = (await res.json()) as IMedication;

          if (fetchAbort.signal.aborted) {
            return;
          }

          setMedication(newData);
          cacheMedication(newData);
          setUpdateAvailable(true);
        } else {
          setLoadFailed(true);
        }
      } catch (ex) {
        if (!fetchAbort.signal.aborted) {
          // tslint:disable-next-line:no-console
          console.log(`Could not get from ${detailEndpoint}/${id}`);
          setLoadFailed(true);
        }
      } finally {
        if (!fetchAbort.signal.aborted) {
          setIsLoading(false);
        }
      }
    }

    func();

    return () => fetchAbort.abort();
  }, [isOnline, id, refresh]);

  return {
    isLoading,
    loadFailed,
    medication,
    updateAvailable
  };
}

export function cacheMedication(medication: IMedication) {
  const localStoreString = localStorage.getItem(medicationsKey);

  const data = localStoreString
    ? (JSON.parse(localStoreString) as Array<Partial<IMedication>>)
    : undefined;

  let newList: Array<Partial<IMedication>>;

  if (data) {
    newList = data.map(m => (m.id === medication.id ? medication : m));
  } else {
    newList = [medication];
  }

  localStorage.setItem(medicationsKey, JSON.stringify(newList));
}
