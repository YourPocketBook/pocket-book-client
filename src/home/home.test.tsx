import { render } from "@testing-library/react";
import React from "react";
import ReactDOM from "react-dom";
import { Home } from ".";
import { useOnlineStatus } from "../hooks/useOnlineStatus";

jest.mock("../hooks/useOnlineStatus");

describe("Home Page", () => {
  beforeEach(() => {
    (useOnlineStatus as jest.Mock).mockReturnValue(true);
  });

  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(<Home />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it("renders correctly", () => {
    const { container } = render(<Home />);

    expect(container).toMatchSnapshot();
  });

  it("renders correctly when offline", () => {
    (useOnlineStatus as jest.Mock).mockReturnValue(false);

    const { container } = render(<Home />);

    expect(container).toMatchSnapshot();
  });
});
