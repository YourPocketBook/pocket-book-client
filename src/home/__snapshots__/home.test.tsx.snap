// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Page renders correctly 1`] = `
<div>
  <div
    class="background"
  >
    <nav
      class="navbar navbar"
    >
      <a
        class="navbar-brand"
      >
        PocketBook
      </a>
    </nav>
    <div
      class="centreContainer container"
    >
      <div
        class="row"
      >
        <div
          class="col-lg-6 offset-lg-3"
        >
          <div
            class="centreBox"
          >
            <div
              class="container"
            >
              <div
                class="row"
              >
                <div
                  class="col"
                >
                  <h1
                    class="text-center"
                  >
                    Pocket Book
                  </h1>
                </div>
              </div>
              <div
                class="row"
              >
                <div
                  class="col"
                >
                  <p>
                    Welcome to Pocket Book. You will need an account to access this site.
                  </p>
                </div>
              </div>
              <div
                class="row"
              >
                <div
                  class="col-sm-4 offset-sm-2"
                >
                  <a
                    class="btn btn-dark-green btn-block"
                    href="/register"
                    id="register-button"
                  >
                    Register
                  </a>
                </div>
                <div
                  class="col-sm-4"
                >
                  <a
                    class="btn btn-dark-green btn-block"
                    href="/login"
                    id="log-in-button"
                  >
                    Log In
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <nav
      class="fixed-bottom footer navbar"
    >
      <span
        class="navbar-text"
      >
        Design Copyright © 
        2020
         Tony Richards.
      </span>
      <ul
        class="nav"
      >
        <li
          class="nav-item"
        >
          <a
            class="nav-link text-dark"
            href="/privacy"
          >
            Privacy Policy
          </a>
        </li>
        <li
          class="nav-item"
        >
          <a
            class="nav-link text-dark"
            href="/tsandcs"
          >
            Terms of Use
          </a>
        </li>
      </ul>
    </nav>
  </div>
</div>
`;

exports[`Home Page renders correctly when offline 1`] = `
<div>
  <div
    class="background"
  >
    <nav
      class="navbar navbar"
    >
      <a
        class="navbar-brand"
      >
        PocketBook
      </a>
    </nav>
    <div
      class="centreContainer container"
    >
      <div
        class="row"
      >
        <div
          class="col-lg-6 offset-lg-3"
        >
          <div
            class="centreBox"
          >
            <div
              class="container"
            >
              <div
                class="alert alert-dark fade"
                role="alert"
              >
                You appear to be offline.
              </div>
              <div
                class="row"
              >
                <div
                  class="col"
                >
                  <h1
                    class="text-center"
                  >
                    Pocket Book
                  </h1>
                </div>
              </div>
              <div
                class="row"
              >
                <div
                  class="col"
                >
                  <p>
                    Welcome to Pocket Book. You will need an account to access this site.
                  </p>
                </div>
              </div>
              <div
                class="row"
              >
                <div
                  class="col-sm-4 offset-sm-2"
                >
                  <a
                    class="btn btn-dark-green btn-block disabled"
                    href="/register"
                    id="register-button"
                  >
                    Register
                  </a>
                </div>
                <div
                  class="col-sm-4"
                >
                  <a
                    class="btn btn-dark-green btn-block disabled"
                    href="/login"
                    id="log-in-button"
                  >
                    Log In
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <nav
      class="fixed-bottom footer navbar"
    >
      <span
        class="navbar-text"
      >
        Design Copyright © 
        2020
         Tony Richards.
      </span>
      <ul
        class="nav"
      >
        <li
          class="nav-item"
        >
          <a
            class="nav-link text-dark"
            href="/privacy"
          >
            Privacy Policy
          </a>
        </li>
        <li
          class="nav-item"
        >
          <a
            class="nav-link text-dark"
            href="/tsandcs"
          >
            Terms of Use
          </a>
        </li>
      </ul>
    </nav>
  </div>
</div>
`;
